src/smallmap_gui.cpp

```diff
@@ -1004,6 +1009,45 @@ class SmallMapWindow : public Window {
 		}
 	}
 
+	/**
+	 * Adds industry names to the smallmap.
+	 * @param dpi the part of the smallmap to be drawn into
+	 */
+	void DrawIndustryNames(const DrawPixelInfo *dpi, const int vertical_padding) const
+	{
+		if (this->map_type != SMT_INDUSTRY) return;
+
+		for (const Industry *i : Industry::Iterate()) {
+			const LegendAndColour &tbl = _legend_from_industries[_industry_to_list_pos[i->type]];
+			if (!tbl.show_on_map) continue;
+
+			/* Industry names blink together with their blobs in the smallmap. */
+			const bool is_blinking = i->type == _smallmap_industry_highlight && !_smallmap_industry_highlight_state;
+			if (is_blinking) continue;
+
+			if (_industry_to_name_string_width[i->type] == 0) {
+				_industry_to_name_string_width[i->type] = GetStringBoundingBox(tbl.legend, FS_SMALL).width;
+			}
+			const uint8_t &legend_text_width = _industry_to_name_string_width[i->type];
```

Comment:
Isn't 255 pixels width a bit too small? If the user selects 5x interface size, that would only be 51 pixels in the normal size. That seems quite narrow for an industry name.

Reply:
@rubidium42 Good catch! The default industries come in just under 255 at 5x:

![screenshot of small map showing industry names at 5x interface scale](https://github.com/OpenTTD/OpenTTD/assets/1967998/50626226-b2c6-4212-bd71-4dc6a666ccec)

![screenshot of small map showing industry name label widths at 5x interface scale](https://github.com/OpenTTD/OpenTTD/assets/1967998/e4181a85-6d64-481e-a0e1-277ba4a61047)

1x for comparison:

<img width="825" alt="screenshot of small map showing industry names at 1x interface scale" src="https://github.com/OpenTTD/OpenTTD/assets/1967998/966feb83-2d55-4816-b49c-62a402c958b3">

![screenshot of small map showing industry name label widths at 1x interface scale](https://github.com/OpenTTD/OpenTTD/assets/1967998/f6e1cd7d-cb38-43ee-a082-bbddc4a4bca7)

This is indeed too tight for other custom industries with longer names. Updated from `uint8_t` to `uint16_t` now.

While investigating this, I found a bug where the cached widths are not updated when changing the interface scale. Updated to reset the cache in `OnInit()` now.

Since we now call `ReInit()` each time we switch the map type https://github.com/OpenTTD/OpenTTD/pull/12770#discussion_r1650212639, this does mean the cached widths are recalculated each time the industry view is selected. The cache would only be used while toggling individual industry legends.

The cache-busting could be more precise by adding a "zoom changed" virtual method from `Window`. However, the added complexity of maintaining a new method/API in the common `Window` class is not worth it for this single use case imo.

https://github.com/OpenTTD/OpenTTD/blob/9936b3728abdde9e8bfc792767753f112981b4d4/src/window.cpp#L3304-L3312